instance/models/instance.py

```diff
@@ -0,0 +1,476 @@
+from instance.repo import clone_configuration_repo
+from instance.models.logging_mixin import LoggerInstanceMixin
+from instance.models.utils import ValidateModelMixin
+
+
+# Constants ###################################################################
+
+PROTOCOL_CHOICES = (
+    ('http', 'HTTP - Unencrypted clear text'),
+    ('https', 'HTTPS - Encrypted'),
+)
+
+gandi = GandiAPI()
+
+
+# Exceptions ##################################################################
+
+class InconsistentInstanceState(Exception):
+    """
+    Indicates that the status of an instance can't be determined
+    """
+    pass
+
+
+# Validators ##################################################################
+
+sha1_validator = RegexValidator(regex='^[0-9a-f]{40}$', message='Full SHA1 hash required')
+
+
+# Models ######################################################################
+
+class Instance(ValidateModelMixin, TimeStampedModel):
+    """
+    Instance - Group of servers running an application made of multiple services
+    """
+    EMPTY = 'empty'
+    NEW = 'new'
+    STARTED = 'started'
+    ACTIVE = 'active'
+    BOOTED = 'booted'
+    PROVISIONED = 'provisioned'
+    REBOOTING = 'rebooting'
+    READY = 'ready'
+    LIVE = 'live'
+    STOPPING = 'stopping'
+    STOPPED = 'stopped'
+    TERMINATING = 'terminating'
+
+    sub_domain = models.CharField(max_length=50)
+    email = models.EmailField(default='contact@example.com')
+    name = models.CharField(max_length=250)
+
+    base_domain = models.CharField(max_length=50, default=settings.INSTANCES_BASE_DOMAIN)
+    protocol = models.CharField(max_length=5, default='http', choices=PROTOCOL_CHOICES)
+
+    last_provisioning_started = models.DateTimeField(blank=True, null=True)
+
+    class Meta:
+        abstract = True
+        unique_together = ('base_domain', 'sub_domain')
+
+    def __str__(self):
+        return '{0.name} ({0.url})'.format(self)
+
+    @property
+    def domain(self):
+        """
+        Instance domain name
+        """
+        return '{0.sub_domain}.{0.base_domain}'.format(self)
+
+    @property
+    def url(self):
+        """
+        Instance URL
+        """
+        return u'{0.protocol}://{0.domain}/'.format(self)
+
+    @property
+    def active_server_set(self):
+        """
+        Returns the subset of `self.server_set` which aren't terminated
+        """
+        return self.server_set.exclude_terminated()
+
+    @property
+    def status(self):
+        """
+        Instance status
+        """
+        active_server_set = self.active_server_set
+        if not active_server_set:
+            return self.EMPTY
+        elif active_server_set.count() > 1:
+            raise InconsistentInstanceState('Multiple servers are active, which is unsupported')
+        else:
+            return active_server_set[0].status
+
+
+# Git #########################################################################
+
+class VersionControlInstanceMixin(models.Model):
```

Comment:
Do these mixins needs to inherit from `models.Model`? You have quite a diamond situation going on here. I think it would be cleaner for the mixins to have no base class unless it's necessary.